include/vcpkg/base/expected.h

```diff
@@ -105,191 +26,328 @@ namespace vcpkg
+        // Otherwise, returns error() put into the same type.
+    private:
+        template<class Test>
+        struct IsThenCompatibleExpected : std::false_type
+        {
+        };
+
+        template<class OtherTy>
+        struct IsThenCompatibleExpected<ExpectedT<OtherTy, S>> : std::true_type
+        {
+        };
+
+    public:
         template<class F, class... Args>
-        std::invoke_result_t<F, const_ref_type, Args&&...> then(F f, Args&&... args) const&
+        std::invoke_result_t<F, const T&, Args...> then(F f, Args&&... args) const&
         {
-            if (has_value())
+            static_assert(IsThenCompatibleExpected<std::invoke_result_t<F, const T&, Args...>>::value,
+                          "then expects f to return an expected with the same error type");
+            if (value_is_error)
             {
-                return std::invoke(f, *m_t.get(), static_cast<Args&&>(args)...);
+                return {m_s, expected_right_tag};
             }
             else
             {
-                return std::invoke_result_t<F, const_ref_type, Args&&...>{m_s};
+                return std::invoke(f, *m_t.get(), static_cast<Args&&>(args)...);
             }
         }
 
         template<class F, class... Args>
-        std::invoke_result_t<F, move_ref_type, Args&&...> then(F f, Args&&... args) &&
+        std::invoke_result_t<F, T&&, Args...> then(F f, Args&&... args) &&
         {
-            if (has_value())
+            static_assert(IsThenCompatibleExpected<std::invoke_result_t<F, T&&, Args...>>::value,
+                          "then expects f to return an expected with the same error type");
+            if (value_is_error)
             {
-                return std::invoke(f, std::move(*m_t.get()), static_cast<Args&&>(args)...);
+                return {m_s, expected_right_tag};
             }
             else
             {
-                return std::invoke_result_t<F, move_ref_type, Args&&...>{std::move(m_s)};
+                return std::invoke(f, std::move(*m_t.get()), static_cast<Args&&>(args)...);
             }
         }
 
     private:
-        template<class, class>
-        friend struct ExpectedT;
-
-        explicit ExpectedT(const ErrorHolder<S>& err) : m_s(err) { }
-        explicit ExpectedT(ErrorHolder<S>&& err) : m_s(static_cast<ErrorHolder<S>&&>(err)) { }
-
         void exit_if_error(const LineInfo& line_info) const
         {
-            if (m_s.has_error())
+            if (value_is_error)
             {
-                msg::write_unlocalized_text_to_stdout(Color::error, Strings::concat(m_s.to_string(), '\n'));
-                Checks::exit_fail(line_info);
+                Checks::exit_with_message(line_info, to_string(error()));
             }
         }
 
-        ErrorHolder<S> m_s;
-        ExpectedHolder<T> m_t;
+        void exit_if_not_error() const
+        {
+            if (!value_is_error)
+            {
+                Checks::unreachable(VCPKG_LINE_INFO);
```

Comment:
In a future change I would like to rename value_or_exit to just value, and add a line_info parameter to error(), but that change is so mechanical and cross-cutting that I want it to be its own review.